### ISSUE TITLE
QT GUI usability problems

### ISSUE THREAD

**Issue by omnidan** [ElektraInitiative/libelektra#2030]
the usability test for Elektra Web also uncovered some problems with the QT GUI:

- [ ] when keys have subkeys, but none of those subkeys have further subkeys, they are shown with a black arrow instead of the usual green one:

![screen shot 2018-05-26 at 20 34 45](https://user-images.githubusercontent.com/668674/40579296-8258606e-6124-11e8-8b2b-b088641f8473.png)

- [ ] in general it is not clear where information appears (e.g. some users didn't see the search results, others clicked on a key and didn't see that something popped up on the right)
- [ ] key values are only shown when hovering over the key or double clicking to get to the edit dialog
- [ ] copy & pasting array keys does not work (creates a key with the same id)
- [ ] it is not possible to edit names of array keys (which makes it impossible to fix the copy & paste issue)
- [x] it is not clear that the box when editing a key contains metakeys (only the button at the bottom "new meta key" hints at it), some users thought it contains subkeys, maybe add a title to this section

**Comment by markus2330** [comment]
Thank you for reporting the problems! I think some clarifications are needed (what was expected, what the users tried to do,...)

Maybe you can also separate usability problems and the actual bug (copy&pasting array keys).

@0003088 please write if it is not clear to you what the individual problems were.

**Comment by markus2330** [comment]
> Do I guess correctly that you tested your GUI with an audience and included the qt-gui as well?

Yes, it was an A/B test with Web-UI and qt-gui. You can read about the details in Daniel's bachelor thesis (to be finished soon, preliminary version is in our private repo).

> Or is the problem that the subkey indicator does not look similar to the expansion icon?

Yes, I also think it is a good idea to have different icons there. Maybe the usability is better if the "Key Name/Key Value" view is also used if non-leaves are selected (up to all keys of a namespace). For irregular-shaped "trees" (especially if exactly one key has subkeys but no other) the right view is nearly always empty and the hints need to be used very often.

> That the key values are only shown by hovering or clicking is true, but I did not have any better idea to solve the problem considering the general design of the viewports.

Maybe reducing the time when the hover text is shown might already help.

> The copying and pasting issue of the arrays seems to be a bug as is the next array related icon, I will see what can be done.

Thank you!

We decided that arrays should have the metadata "array" containing the last index (see #182). This should make detection of arrays easier. It would be nice to see support for it in the qt-gui.

**Comment by markus2330** [comment]
@0003088 Maybe we can here also have some improvements (next to #2405)

**Comment by 0003088** [comment]
> Maybe reducing the time when the hover text is shown might already help.

The interval is set for 1 second atm, so set it to 0.5s? Or maybe even  add an option for the user to select the time for herself?

> Maybe the usability is better if the "Key Name/Key Value" view is also used if non-leaves are selected (up to all keys of a namespace). For irregular-shaped "trees" (especially if exactly one key has subkeys but no other) the right view is nearly always empty and the hints (hover text) need to be used very often.

Maybe add a checkbox in the settings as well? Called "always show key name/key value" or something like that?

**Comment by markus2330** [comment]
> The interval is set for 1 second atm, so set it to 0.5s?

Yes, sounds good.

> Or maybe even add an option for the user to select the time for herself?

I think we should choose it, as we have the responsibility to provide usable programs.

> Maybe add a checkbox in the settings as well? Called "always show key name/key value" or something like that?

If there is no strong reason against it, I would also simply always show (without any setting).

**Comment by markus2330** [comment]
@0003088  Thank you for working on this!

**Comment by 0003088** [comment]
@markus2330 You are welcome, but I cannot promise any timeframe when this will be done. Sorry for the unintended closing of this issue. I should avoid the term 'fix' I suppose ^^

**Comment by markus2330** [comment]
No, it is only that you directly pushed into master and because of that the "fixes ..." in comments immediately triggers to close the issues. If you create a PR, this will not happen (only after the PR was merged).

Please do not forget to write release notes. (doc/news/_preparation_next_release.md)

**Comment by 0003088** [comment]
Ugh, you're right, I didn't work out of my fork! Do you want to reverse the commits?

**Comment by markus2330** [comment]
If they are okay, you can leave them. Please only add a note about them in the release notes.

**Comment by 0003088** [comment]
It seems to be fine on my machine. I'm not worried about the timer, but I fixed some layout errors and I do not know how older versions of Qt will react to these changes, since I've never seen the errors before (there's no issue for this I just fixed them)

**Comment by markus2330** [comment]
@darddan ping

**Comment by markus2330** [comment]
@0003088 I unassigned you, as @darddan will now work on it. We will integrate the qt-gui more nicely into KDE.

**Comment by 0003088** [comment]
Sounds nice, looking forward to see improvements.

**Comment by markus2330** [comment]
@darddan did you have a look at these issues? Could you reproduce them?

**Comment by darddan** [comment]
The issues were all there when I last checked. Unfortunately because of my health condition my time in front of the computer is very limited and didn't have any time to work on these issues. My current long-time-open-task is the migration to a newer version of `QML` and using the "official" `TreeView` (but I can't tell how close I am at finishing that either) and that one doesn't affect the problems stated in this task.

**Comment by markus2330** [comment]
Beautiful to read from you again, I wish you to recover soon! :heart: 

Good plan, would be amazing to use the official TreeView.

**Comment by 0003088** [comment]
I'm excited to see somebody is planning to use the official tree-view. I myself tried to migrate the qt-gui to the new view a few years ago, when the qml view was released, but failed unfortunately (if anybody is interested, the code is in https://github.com/0003088/libelektra-qt-gui-test/tree/master/src/tools/qt-gui). AFAIR the problem was the proxy model. As long as the nodes where preserved in the structure, e.g. just renamed, it would work fine, but as soon as I started to add and remove nodes I would run into problems I could not solve. I'm courious to see how it can be done.
